botorch/models/utils/inducing_point_allocators.py

```diff
@@ -0,0 +1,388 @@
+                        max_length=num_inducing,
+                        quality_scores=quality_scores,
+                    )
+                )
+            inducing_points = torch.stack(inducing_points).view(
+                *input_batch_shape, num_inducing, -1
+            )
+
+        return inducing_points
+
+
+class QualityFunction(ABC):
+    """
+    A function that scores inputs with respect
+    to a specific criterion.
+    """
+
+    @abstractmethod
+    def __call__(inputs: Tensor) -> Tensor:  # [n, d] -> [n]
+        """
+        Args:
+            inputs: inputs (of shape n x d)
+        Returns:
+            A tensor of quality scores for each input, of shape [n]
+        """
+
+        pass
+
+
+class UnitQualityFunction(QualityFunction):
+    """
+    A function returning ones for each element. Using this quality function
+    for inducing point allocation corresponds to allocating inducing points
+    with the sole aim of minimizing predictive variance, i.e. the approach
+    of [burt202svgp]_.
+    """
+
+    @torch.no_grad()
+    def __call__(self, inputs: Tensor) -> Tensor:  # [n, d]-> [n]
+        """
+        Args:
+            inputs: inputs (of shape n x d)
+        Returns:
+            A tensor of ones for each input, of shape [n]
+        """
+        return torch.ones([inputs.shape[0]], dtype=inputs.dtype)
+
+
+class ExpectedImprovementQualityFunction(QualityFunction):
+    """
+    A function measuring the quality of input points as their expected
+    improvement with respect to a conservative baseline. Expectations
+    are according to the model from the previous BO step. See [moss2023ipa]_
+    for details and justification.
+    """
+
+    def __init__(self, model: Model, maximize: bool):
+        r"""
+        Args:
+            model: The model fitted during the previous BO step. For now, this
+                must be a single task model (i.e. num_outputs=1).
+            maximize: Set True if we are performing function maximization, else
+                set False.
+        """
+        if model.num_outputs != 1:
+            raise NotImplementedError(
+                "Multi-output models are currently not supported. "
+            )
+        self._model = model
+        self._maximize = maximize
+
+    @torch.no_grad()
+    def __call__(self, inputs: Tensor) -> Tensor:  # [n, d] -> [n]
+        """
+        Args:
+            inputs: inputs (of shape n x d)
+        Returns:
+            A tensor of quality scores for each input, of shape [n]
+        """
+
+        posterior = self._model.posterior(inputs)
+        mean = posterior.mean.squeeze(-2).squeeze(-1)  # removing redundant dimensions
+        sigma = posterior.variance.clamp_min(1e-12).sqrt().view(mean.shape)
+
+        best_f = torch.max(mean) if self._maximize else torch.min(mean)
+        u = (mean - best_f) / sigma if self._maximize else -(mean - best_f) / sigma
+        return sigma * (phi(u) + u * Phi(u))
```

Comment:
I wonder if we can reuse some of the numerics from @SebastianAment's recent work here (see e.g. here: https://github.com/pytorch/botorch/blob/main/botorch/acquisition/analytic.py#L174-L176). I guess this isn't logEI here and gradients aren't really of interest here, but could still make sense to leverage the EI computation from somewhere rather than re-implementing it here.

Reply:
I had lots of circular import issues. Do you have a specific request for the function you want me to use?

Reply:
There's the [`_log_ei_helper`](https://github.com/pytorch/botorch/blob/a8efd76b43c19982d5d0b79f14912b27094b5d93/botorch/acquisition/analytic.py#L836), which computes `log(phi(u) + u * Phi(u))` in a numerically accurate and stable fashion. Even if you don't need gradients, it could be worth returning `LogEI` here instead because it will be able to differentiate (no pun intended) between points whose naïve `EI` is numerically - but not mathematically - zero.

There's also [`_scaled_improvement`](https://github.com/pytorch/botorch/blob/a8efd76b43c19982d5d0b79f14912b27094b5d93/botorch/acquisition/analytic.py#L821), which computes `u`, and [`_ei_helper`](https://github.com/pytorch/botorch/blob/a8efd76b43c19982d5d0b79f14912b27094b5d93/botorch/acquisition/analytic.py#L829), which computes `phi(u) + u * Phi(u)` naïvely. While it wouldn't change the functionality, using the existing helpers would allow future improvements to translate automatically to this code.

Reply:
Hmm I think it's ok for now; this isn't that crazy. Once we merge all of @SebastianAment's stuff in we can probably take another more global look at how to reuse those in other places.

Reply:
If you get import errors for the helpers, it could make sense to move `_log_ei_helper` / `_ei_helper` to the utils. I can do that and adjust the other dependencies, just let me know!

Reply:
yeah that would be great pls @SebastianAment